### ISSUE TITLE
MemberTypes not found, yet the immediate window can find the type.

### ISSUE THREAD

**Issue by NullVoxPopuli** [dotnet/corefx#4670]
I'm not sure if this is a runtime mismatch of something

Here is my method:

```
   private static object ValueFor(object obj, MemberInfo member)
        {
            object value = null;

            if (member != null)
            {
                switch (member.MemberType)
                {
                    case MemberTypes.Field:
                        value = ((FieldInfo)member).GetValue(obj);
                        break;
                    case MemberTypes.Property:
                        value = ((PropertyInfo)member).GetValue(obj, null);
                        break;
                    case MemberTypes.Method:
                        break;

                    default:
                        throw new System.ArgumentOutOfRangeException();
                }
            }

            return value;
        }
```

And then in my immediate window:

```
?member.MemberType
Property
?member.MemberType.GetType()
{Name = "MemberTypes" FullName = "System.Reflection.MemberTypes"}
    Assembly: {mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e}
```

My dependencies in project.json

```
{
    "dependencies": {
        "EntityFramework.MicrosoftSqlServer": "(7.0.0-rc1-final,]",
        "EntityFramework.MicrosoftSqlServer.Design": "(7.0.0-rc1-final,]",
        "EntityFramework.Commands": "(7.0.0-rc1-final,]",
        "System.Reflection": "4.1.0-beta-*",
        "Microsoft.Extensions.PlatformAbstractions": "(1.0.0-rc1-final,]"
    },
   ...
```

So there looks to be a version mismatch of System.Reflection.  What's up with that?

I'm debugging in Visual Studio 2015.

**Comment by NullVoxPopuli** [comment]
@pallavit or @joshfree do you know what the problem is? or if there is a work around?

**Comment by stephentoub** [comment]
@NullVoxPopuli, it's not clear to me from your description what the actual problem is.  Can you elaborate?  What behavior are you seeing at run time that's unexpected?

**Comment by NullVoxPopuli** [comment]
the problem is actually at compile time. 

`dnx run` gives me these errors (if I don't comment out the references to MemberType(s) before running - for debugging)

```
Severity	Code	Description	Project	File	Line
Error	CS1061	'MemberInfo' does not contain a definition for 'MemberType' and no extension method 'MemberType' accepting a first argument of type 'MemberInfo' could be found (are you missing a using directive or an assembly reference?)	USE.DNX Core 5.0	D:\ClientDevelopment\AllisonTransmission\USE\Extensions\ObjectExtensions\Methods.cs	131
Error	CS0103	The name 'MemberTypes' does not exist in the current context	USE.DNX Core 5.0	D:\ClientDevelopment\AllisonTransmission\USE\Extensions\ObjectExtensions\Methods.cs	133
```

**Comment by stephentoub** [comment]
Ah, I see.  The discrepancy you're seeing isn't a version mismatch, it's that the member you want isn't exposed in the contract (the contracts, or reference assemblies, often represent a smaller surface area than what's actually available in the underlying implementation on a given platform).  @terrajobst, @weshaggard, or @nguerrera could likely speak to why it's not part of the surface area and what the recommended alternative is.

**Comment by weshaggard** [comment]
@NullVoxPopuli we are going to be taking a look at the missing surface area and trying to determine what to bring back these will certainly be considered.

As for a workaround to get you unblocked right now the best you can do is a cast to determine the type. Something along the lines:

```
FieldInfo field = member as FieldInfo;
if (field != null)
   return field.GetValue(obj);
PropertyInfo property member as PropertyInfo;
if (property != null)
   return property.GetValue(obj, null);
```

**Comment by NullVoxPopuli** [comment]
Thanks for the workaround. I look forward to an alternative :-)

**Comment by weshaggard** [comment]
@NullVoxPopuli I personally prefer the as casts over the MemberTypes case statement for this type of work anyway, but that is just my own personal preference. 

Given you are unblocked now and the general tracking of the reflection API gaps is already tracked I'm going to close this issue now. Thanks for reporting your issue @NullVoxPopuli.
